src/main/scala/devices/debug/Debug.scala

```diff
@@ -1159,3 +1418,89 @@ class DMIToTL(implicit p: Parameters) extends LazyModule {
 
   }
 }
+
+class SBToTL(implicit p: Parameters) extends LazyModule {
+
+  val node = TLClientNode(Seq(TLClientPortParameters(Seq(TLClientParameters("sba")))))
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val rdEn         = Bool(INPUT)
+      val wrEn         = Bool(INPUT)
+      val addrIn       = UInt(INPUT, 128) // TODO: Parameterize these widths
+      val sizeIn       = UInt(INPUT, 128)
+      val dataIn       = UInt(INPUT, 128)
+      val dmactive     = Bool(INPUT)
+      val dataOutValid = Bool(OUTPUT)
+      val dataOut      = UInt(OUTPUT, 128)
+      val sbStateOut   = UInt(OUTPUT, log2Ceil(SystemBusAccessState.maxId))
+    })
+
+    import uncore.devices._
+    import uncore.devices.DMI_RegAddrs._
+ 
+    val (tl, edge) = node.out(0)
+ 
+    // --- Drive payloads on bus to TileLink ---
+    val src  = Wire(init = 0.U)
+    val (_,  gbits) = edge.Get(src, io.addrIn, io.sizeIn)
+    val (_, pfbits) = edge.Put(src, io.addrIn, io.sizeIn, io.dataIn)
+    val (_,  nbits) = edge.Put(src, io.addrIn, io.sizeIn, data=0.U, mask=0.U)
+    val sbState = Reg(init = 0.U)
+ 
+    io.sbStateOut := sbState
+    when       (sbState === SystemBusAccessState.SBReadRequest.id.U) { tl.a.bits := pfbits
+    }.elsewhen (sbState === SystemBusAccessState.SBWriteRequest.id.U){ tl.a.bits :=  gbits
+    }.otherwise                                                 {      tl.a.bits :=  nbits
+    }
+ 
+    // --- State Machine to interface with TileLink ---
+    val requestValid = tl.a.valid
+    val requestReady = tl.a.ready
+ 
+    val responseValid = tl.d.valid
+    val responseReady = tl.d.ready
+ 
+    when(~io.dmactive){
+      sbState := SystemBusAccessState.Idle.id.U
+    }.elsewhen (sbState === SystemBusAccessState.Idle.id.U){
+      sbState := Mux(io.rdEn, SystemBusAccessState.SBReadRequest.id.U,
+                 Mux(io.wrEn, SystemBusAccessState.SBWriteRequest.id.U,sbState))
+    }.elsewhen (sbState === SystemBusAccessState.SBReadRequest.id.U){
+      sbState := Mux(requestValid && requestReady, SystemBusAccessState.SBReadResponse.id.U, sbState) 
```

Comment:
this seems sort of circular logic. requestValid depends on tl.a.valid which is depending on SystmBusAccessState == SBReadRequest, so why include it in this check again? Basically as soon as you see the `ready` you can progress, right?

Reply:
Yeah, I see how it reads that way, and I could move to the Response state when I see "ready" from TileLink, I just wanted to be explicit that the request happens when valid && ready. There should be no circular dependency since sbState is a register